### DISCUSSION TITLE
tutorials/intro-typescript/08-codegen

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#101] [course feedback]
# tutorials/intro-typescript/08-codegen

https://www.apollographql.com/tutorials/intro-typescript/08-codegen

**Comment by YuanRuQian** [comment]
> We have everything we need to run the codegen command. Open up a terminal in the server directory and run the following command.

Somehow I could not find the `server` directory. And running `npm run generate` in the root folder will cause `Error: Unable to find Codegen config file! `. So my alternative approach is to change the `generate` setup to `graphql-codegen --config ./src/codegen.ts`.

**Reply by lizhennessy** [reply]
Hi there @YuanRuQian ! Thanks for pointing this out; the `server` directory reference is definitely a typo. I'm on the fix now! We should be able to run the command in the root of our project directory. Did you place your `codegen.ts` in the root, or inside the `src` folder? It should be in the root instead. :)

**Reply by YuanRuQian** [reply]
Ah, I mistakenly placed the `codegen.ts` file in the `src` folder! After relocating it to the root folder, the `npm run generate` works. Thank you!

**Comment by moishinetzer** [comment]
A lot of the typings aren't being enforced properly because `strict: true` is missing from the `tsconfig.ts` which is why the resolvers accept `null` as a return type for required arguments

**Comment by moishinetzer** [comment]
```tsx
import { Resolvers } from "./types";

export const resolvers: Resolvers = {
  Query: {
    featuredPlaylists: (_, __, { dataSources }) => {
      return dataSources.spotifyAPI.getFeaturedPlaylists();
    },
  },
};
```

Is it the case that you get more accurate types by using the `satisfies` instead?

```tsx
import { Resolvers } from "./types";

export const resolvers = {
  Query: {
    featuredPlaylists: (_, __, { dataSources }) => {
      return dataSources.spotifyAPI.getFeaturedPlaylists();
    },
  },
} satisfies Resolvers
```

**Reply by BlenderDude** [reply]
@moishinetzer 

`const resolvers: Resolvers` strictly defines the exact keys and values, so the only difference between the two is the final type. `const resolvers: Resolvers` means the variable will be of type `Resolvers`. For satisfies, the final type is as follows:
```ts
// hovering resolvers shows:
{
  Query: {
    featuredPlaylists: (_, __, {dataSources}:{dataSources: DataSources}) => Playlist[]
  }
}
```
which satisfies `Resolvers`, but is _exactly_ what you wrote. The `Resolvers` type includes a ton of optional fields to describe every _possible_ type and field.

Unless there is a type-system library or other tool using the explicit output typing from `satisfies`, it is better to use `const resolver: Resolver` so you know elsewhere it is a Resolver map. Otherwise using it (in this case) decreases the clarity, type-hinting, and developer experience of your code base.

To help solidify this concept, here is an example of where the slimmed output of `satisfies` comes in handy. Let's define a dictionary type like this:
```ts
type Dictionary = Record<string, any>
```
Simple enough -- anything that is a dictionary must be an object where the keys are any string. Let's also define a function `dictionaryKeys` that takes a dictionary and returns an array of its keys. This requires a generic which is a bit advanced, but the point here is that we need to know _exactly_ what dictionary was passed to the function:

```ts
function dictionaryKeys<D extends Dictionary>(dictionary: D): Array<keyof D> {
  return Object.keys(dictionary);
}
```

Great! Now lets create two dictionaries, defined on the variable and satisfies:

```ts
const dictionaryDefined: Dictionary = {
  foo: "bar",
  bar: 1,
  baz: true,
};

const dictionarySatisfies = {
  foo: "bar",
  bar: 1,
  baz: true,
} satisfies Dictionary;
```

Now if we use our `dictionaryKeys` on both to see the difference

```ts
dictionaryKeys(dictionaryDefined) // -> Array<string>

dictionaryKeys(dictionarySatisfies) // -> Array<"foo" | "bar" | "baz">
```

Note how with the defined object, we _erased_ our strict information about what exact keys exist as `Record<string, any>` states our keys are "string" but that is it. When using `satisfies`, we instead keep the type information about our _specific_ object and Typescript can know the keys are `"foo" | "bar" | "baz"`.

**Comment by moishinetzer** [comment]
It appears the layout of the discussion bot is broken on this page

<img width="1710" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/11411486/c90a964f-5607-4838-85e8-07721bcdb98e">

**Reply by lizhennessy** [reply]
Oof! Thanks for the heads-up @moishinetzer !

**Reply by AlexRebula** [reply]
This issue seems not to be resolved yet.

**Comment by moishinetzer** [comment]
Sorry, lots of feedback and comments but I haven't enjoyed a course like this in a long time! Incredible stuff.

**Comment by moishinetzer** [comment]
<img width="361" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/11411486/25c1b61c-b716-4b9a-93e6-e23b760d2642">

When there are no practice questions the sidebar does not show as completed when clicking "Next"

**Reply by lizhennessy** [reply]
Thanks for letting us know @moishinetzer ! I'll take a look. 😊

**Comment by Cavein254** [comment]
I notice a type issue in the `return` portion. `Type 'Playlist | undefined[]' is not assignable to type 'Playlist[]'.`

**Reply by Cavein254** [reply]
```js
import { Playlist } from "../types";
import { RESTDataSource } from "@apollo/datasource-rest";

export class Spotify extends RESTDataSource {
  baseURL = "https://spotify-demo-api-fe224840a08c.herokuapp.com/v1/";
  async getFeaturedPlaylists(): Promise<Playlist[]> {
    const response = await this.get<{ playlists: { items: Playlist } }>(
      "browse/featured-playlists"
    );
    return response?.playlists?.items ?? [];
  }
}
```

**Comment by AskYous** [comment]
Codegen, Schema, Types, Resolvers...

It would be nice to summarize all these at the end of this lesson.

**Reply by mabuyo** [reply]
Thanks for the feedback, @AskYous - the "Key Takeaways" section could use a couple more helpful bullet points! We'll log this feedback for a future iteration 📝

**Comment by mark-P0** [comment]
Small typo in the **Overview** section: `SpotifyApi` should be `SpotifyAPI` I believe

Also an opinion, from what I've understood so far: assigning the `Playlist` type from the GraphQL schema as the **response** type of the Spotify REST API seems concerning

**Reply by lizhennessy** [reply]
Hi there @mark-P0 ! You're spot on: the REST API response _is_ different (in significant ways!) from our GraphQL `Playlist` type. As you advance through the course, we'll tackle these differences more thoroughly by introducing mappers into the codegen process. (Hint: Check out [lesson 13](https://www.apollographql.com/tutorials/intro-typescript/13-using-models) for a sneak peek!)

And thanks for the typo catch! We're on it.
